### PR TITLE
planner: add heuristic rules for index selection

### LINKED ISSUE #26319: Adding new heuristic rules

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

close #26319

### What is changed and how it works?

What's Changed & How it Works:

In `(*DataSource).DeriveStats`, we try heuristic rules to select the path from possible access paths. Once some path matches heuristics, we remove other paths from `ds.possibleAccessPaths`, let the selected path remain and add a NOTE level warning to indicate the path matches heuristics.
1. If the query ranges of a unique index are all points and it only needs a single scan, then just use the unique index.
2. If the query ranges of a unique index are all points but it needs a double scan, we collect it into `uniqueIdxsWithDoubleScan`. For the one in `uniqueIdxsWithDoubleScan` with the minimal number of ranges, we record it as `uniqueBest`.
3. If the query ranges of an index are not all points but it only needs a single scan, we collect it into `singleScanIdxs`. For each index in `singleScanIdxs`, if it is better than some index in `uniqueIdxsWithDoubleScan`(there is a subset relation between two `accessCondsColSet`), we call it a refined index. We select the refined index with the minimal number of ranges as `refinedBest`.
4. We compare `uniqueBest` and `refinedBest` and choose the better one.

### Check List 

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [x] Performance regression: Consumes more CPU
- [x] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 
```
add heuristic rules for index selection
```

### CODE REVIEW

File: planner/core/stats.go [hunk 1]
```diff
@@ -280,30 +283,121 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo, selfSchema *
...
+			path.IsSingleScan = ds.isCoveringIndex(ds.schema.Columns, path.FullIdxCols, path.FullIdxColLens, ds.tableInfo)
 		}
-		noIntervalRanges := ds.deriveIndexPathStats(path, ds.pushedDownConds, false)
-		// If we have empty range, or point range on unique index, just remove other possible paths.
-		if (noIntervalRanges && path.Index.Unique) || len(path.Ranges) == 0 {
-			ds.possibleAccessPaths[0] = path
-			ds.possibleAccessPaths = ds.possibleAccessPaths[:1]
-			ds.ctx.GetSessionVars().StmtCtx.OptimDependOnMutableConst = true
+		// Try some heuristic rules to select access path.
+		if len(path.Ranges) == 0 {
```
Comment (qw4990) [inline planner/core/stats.go:307]: When can this happen?

File: planner/core/stats.go [hunk 2]
```diff
@@ -280,30 +283,121 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo, selfSchema *
...
+			selected = path
 			break
 		}
+		if path.OnlyPointRange(ds.SCtx().GetSessionVars().StmtCtx) {
+			if path.IsTablePath() || path.Index.Unique {
+				if path.IsSingleScan {
+					selected = path
```
Comment (qw4990) [inline planner/core/stats.go:314]: What if there are multiple indexes that can hit this rule? Can we pick one of them steadily?

**Review by qw4990 — COMMENTED** [review]
The function `DataSource.DeriveStats` is too huge. It's better to split it into multiple functions, and then put all logic about heuristic rules in one of them?

File: planner/core/stats.go (hunk 1 above)
Comment (xuyifangreeneyes) [reply planner/core/stats.go:307]: When the condition is always false. For example,
https://github.com/pingcap/tidb/blob/853c41e3fbda3b60ca689dd34afce2e187edb000/util/ranger/detacher.go#L272-L273

File: planner/core/stats.go (hunk 2 above)
Comment (xuyifangreeneyes) [reply planner/core/stats.go:314]: The order of `ds.possibleAccessPaths` is steady, so we can pick one of them steadily.

Comment (qw4990) [reply planner/core/stats.go:289]: 🆒

File: planner/core/stats.go [hunk 3]
```diff
@@ -255,6 +258,125 @@ func (ds *DataSource) deriveStatsByFilter(conds expression.CNFExprs, filledPaths
...
+			if uniqueBest == nil || len(uniqueIdx.Ranges) < len(uniqueBest.Ranges) {
+				uniqueBest = uniqueIdx
+			}
+		}
+		// `uniqueBest` may not always be the best.
+		// ```
+		// create table t(a int, b int, c int, unique index idx_b(b), unique index idx_b_c(b, c));
+		// select b, c from t where b = 5 and c > 10;
+		// ```
+		// In the case, `uniqueBest` is `idx_b`. However, `idx_b_c` is better than `idx_b_c`.
```
Comment (qw4990) [inline planner/core/stats.go:312]:
```suggestion
		// In the case, `uniqueBest` is `idx_b`. However, `idx_b_c` is better than `idx_b`.
```

File: planner/core/stats.go [hunk 4]
```diff
@@ -255,6 +258,125 @@ func (ds *DataSource) deriveStatsByFilter(conds expression.CNFExprs, filledPaths
...
+		for _, uniqueIdx := range uniqueIdxsWithDoubleScan {
+			uniqueIdxColumnSets = append(uniqueIdxColumnSets, expression.ExtractColumnSet(uniqueIdx.AccessConds))
+			// Find the unique index with the minimal number of ranges as `uniqueBest`.
+			if uniqueBest == nil || len(uniqueIdx.Ranges) < len(uniqueBest.Ranges) {
+				uniqueBest = uniqueIdx
+			}
+		}
+		// `uniqueBest` may not always be the best.
+		// ```
+		// create table t(a int, b int, c int, unique index idx_b(b), unique index idx_b_c(b, c));
```
Comment (qw4990) [inline planner/core/stats.go:309]: To make this example more clear, can we remove the unique attribution on `idx_b_c`?

Comment (xuyifangreeneyes) [reply planner/core/stats.go]: Done.

File: planner/core/integration_test.go [hunk 5]
```diff
@@ -4152,3 +4152,28 @@ func (s *testIntegrationSuite) TestIssue26559(c *C) {
 	tk.MustExec("insert into t values('2020-07-29 09:07:01', '2020-07-27 16:57:36');")
 	tk.MustQuery("select greatest(a, b) from t union select null;").Sort().Check(testkit.Rows("2020-07-29 09:07:01", "<nil>"))
 }
+
+func (s *testIntegrationSuite) TestHeuristicIndexSelection(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t1, t2")
+	tk.MustExec("create table t1(a int, b int, c int, d int, e int, f int, g int, primary key (a), unique key c_d_e (c, d, e), unique key f (f), unique key f_g (f, g), key g (g))")
+	tk.MustExec("create table t2(a int, b int, c int, d int, unique index idx_a (a), unique index idx_b_c (b, c), unique index idx_b_c_a_d (b, c, a, d))")
```
Comment (qw4990) [inline planner/core/integration_test.go:4161]: Please add some cases about ClusteredIndex.

**Review by qw4990 — COMMENTED** [review]
LGTM

**Review by qw4990 — APPROVED** [review]

**Review by winoros — APPROVED** [review]

**Pull request merged**